### PR TITLE
Add fallback-aware device checking for MPS operations

### PR DESCRIPTION
**Pull request by leochlon**
- Implement FallbackDetector system for context-aware device validation
- Add DeviceCheckType.FallbackAware enum for granular device check control
- Update codegen to emit fallback-aware device checks with MPS support
- Annotate ops with device_check: FallbackAware for legitimate CPU/MPS mixing
- Fixes device compatibility crashes in MPS fallback scenarios

Fixes https://github.com/pytorch/pytorch/issues/142048

### CODE REVIEW

**Comment by leochlon** [comment]
@kulinseth @albanD @malfet @lezcano  Could you please review this MPS device checking improvement? 

Related to issue #142048 building on an earlier fix https://github.com/pytorch/pytorch/pull/142477

**Comment by malfet** [comment]
@leochlon and what is your test plan?

**Comment by leochlon** [comment]
@pytorch-bot label "ciflow/mps"

**Comment by leochlon** [comment]
@malfet thanks for replying!

this is my test plan

1. verify FallbackDetector is compiled into PyTorch
python3 -c "import torch; print('FallbackDetector available:', hasattr(torch._C._dispatch, 'FallbackDetector') if hasattr(torch._C, '_dispatch') else 'Checking c10...')"

2. run our specific MPS regression tests
python3 run_mps_tests.py

3. run our fallback device check tests
python3 -m pytest test/test_fallback_device_checks.py -v

4. run our MPS-specific fallback tests  
python3 -m pytest test/test_mps_fallback_device_checks.py -v


5. verify earlier specific failing CI tests now pass
python3 -m pytest test/test_mps.py::TestLogical::test_isin_asserts -v

@pytorchbot label "ciflow/mps"

**Comment by leochlon** [comment]
@lezcano @kulinseth @malfet @albanD  CPU/MPS 

as of recent commit: 

If the operation (such as linalg_solve_triangular) is annotated as FallbackAware (which it now is), and you mix CPU and MPS tensors:

If fallback is enabled (by setting the environment variable PYTORCH_ENABLE_MPS_FALLBACK=1), the operation will run using the CPU implementation. The tensors will be moved as needed, and the operation will succeed, with a warning indicating that fallback was used.

If fallback is NOT enabled: The operation will still attempt to fallback if it's in a fallback context—such as when the operation is registered for MPS fallback. If not in a fallback context and fallback is not enabled, you will get a error message explaining that CPU/MPS mixing is not allowed for this operation unless fallback is enabled.

For operations that are NOT fallback-aware, CPU/MPS mixing will still result in a device mismatch error as before.

The result tensor's device will depend on the fallback behavior. For linalg_solve_triangular, if fallback occurs, the result will usually be on the CPU. 

this fixes https://github.com/pytorch/pytorch/issues/142048 test plan is running when I build locally , can we run CI please?

**Comment by lezcano** [comment]
The previous issue already had an open PR https://github.com/pytorch/pytorch/pull/155165. Why this new PR?

**Comment by lezcano** [comment]
Also, I'm very tempted to think that this PR is also AI generated without proper further checks (see the emojis), similar to https://github.com/pytorch/pytorch/pull/156247

**Comment by leochlon** [comment]
@lezcano sorry Im still new to contributing trying to understand how everything works which is why I picked this as a "good first issue", I thought that hacking about with someone else's PR would be the equivalent of ruining their work or even plagiarism.

**Comment by leochlon** [comment]
@lezcano I use AI to lint, clean up and document the code to make it easier to follow for people. there's a test plan I included with tests that are passing after I spent hours building it on my macbook. why is this an issue?

**Comment by leochlon** [comment]
@lezcano 

Accusing someone's work of being "AI generated" based on the presence of emojis is both unfounded and unprofessional.
Yes, I used AI tools to help refine documentation and ensure comprehensive test coverage - which is a legitimate and increasingly common practice in modern development. However, the core logic, architecture decisions, and problem-solving were entirely my own work, representing hours of genuine effort.

This follows your previous comment calling my work "slop" on #156247. Making baseless accusations and dismissive remarks about contributors' efforts is exactly the kind of behavior that makes communities toxic and unwelcoming.
If you have legitimate technical concerns about the implementation, I'm happy to address them. But I expect the same professional courtesy I extend to others. Personal attacks and unfounded accusations have no place in constructive code review.

I've put significant time and effort into this contribution, and I'd appreciate if we could focus on the technical merits rather than speculation about my development process.

**Comment by albanD** [comment]
Hey @leochlon 
Thank you for your time trying to contribute here!
We appreciate the time you've put in trying to get PRs ready to fix this issue.

It is also a lot of time on our end to understand and review all these PRs and so you would hopefully understand that PRs that are very far from what is done in the repo, don't follow common practices there, don't have easily understandable context in their description or are different from what is discussed in the issue is prone to waste time on our side.

Please refer to https://github.com/pytorch/pytorch/wiki/The-Ultimate-Guide-to-PyTorch-Contributions for all the details on how to contribute and the workflow to follow if you want to do more than what is discussed in the issue.


For this particular PR, the extension to handle fallback case is a significant enough extension that it should be discussed on the issue before implementing it to make sure we're on the same page on how to do it before you spend a lot of time implementing it (see the details in the guide above).
Similarly, as you can see in https://github.com/pytorch/pytorch/pull/142477 we have a lot of test scaffolding we re-use to improve reliability and coverage of our tests. 

In general in PyTorch, we prefer multiple small and self-contained PRs as opposed to large PRs with many things in them.
To move forward on this one, I would suggest a PR similar to https://github.com/pytorch/pytorch/pull/142477 that only fixes what is discussed inhttps://github.com/pytorch/pytorch/issues/142048. And if you're interested in fixing it for the fallback case as well, open a separate issue where we can discuss what's the best way to do that piece.

> I thought that hacking about with someone else's PR would be the equivalent of ruining their work or even plagiarism.

If the previous author doesn't have time or interest in getting their original PR across the finish line, it is ok to build upon what they've done to land the feature. This is what open source is about afterall :)
